### ISSUE TITLE
It may be a bug that "lradj=constant" is not actually a strictly constant learning-rate schedule.

### ISSUE THREAD

**Issue by oneUTer** [Ruifeng-Tan/BatteryLife#29]
Hello, Professor Tan. Thank you for your excellent work, making this project publicly available.  While reading the code and running some experiments, I observed an unexpected behavior.

# It may be a bug that "lradj=constant" is not actually a strictly constant learning-rate schedule.

After investigating the problem, I believe the abnormal performance of the current **CPGRU** setup is **unlikely to be caused by an incorrect model implementation or by loading the wrong dataset**. A more likely explanation is that the **actual learning-rate behavior** in `BatteryLife-main` is inconsistent with the intuitive meaning of `lradj=constant`.

---

## 1. Main Conclusion

In `BatteryLife-main`, the training script specifies:

- `CPGRU.sh`

```bash
lradj=constant
```

However, during execution, the learning rate is **not truly constant**.

More precisely, the current implementation behaves as follows:

> **The optimizer is first affected by `OneCycleLR` during scheduler initialization, and is later pulled back to `args.learning_rate` by the `constant` branch.**

Therefore, the behavior is not a pure constant schedule. It is better described as:

> **a pseudo-constant schedule influenced by `OneCycleLR`**

---

### Where the parameter comes from
`lradj=constant` is specified in:

- `train_eval_scripts/CPGRU.sh`

### Files that actually implement the learning-rate behavior
1. `run_main.py`
2. `utils/tools.py`

### Key logic in `run_main.py`

The relevant logic is:

```python
if args.lradj == 'COS':
    scheduler = CosineAnnealingLR(...)
else:
    scheduler = OneCycleLR(...)
```

This means:

> **As long as `lradj` is not `COS`, the code still creates `OneCycleLR`.**

So even when the script specifies `lradj=constant`, `OneCycleLR` is still instantiated.

### Key logic in `tools.py`

The relevant branch is:

```python
elif args.lradj == 'constant':
    lr_adjust = {epoch: args.learning_rate}
```

and later:

```python
for param_group in optimizer.param_groups:
    param_group['lr'] = lr
```

This implies:

> **At the end of the epoch, the learning rate is explicitly reset to `args.learning_rate`.**

---

## 3. Evidence 2: `OneCycleLR` Changes the Learning Rate at Initialization

One subtle but important point is easy to miss:

Although `scheduler.step()` is only explicitly called when `lradj == 'TST'`, **`OneCycleLR` already changes the optimizer's current learning rate when the scheduler is created**.

I verified this with a minimal example in the current environment:

```python
import torch
from torch import optim
from torch.optim import lr_scheduler

m = torch.nn.Linear(2, 1)
opt = optim.Adam(m.parameters(), lr=1e-3)
print('before', opt.param_groups[0]['lr'])

sch = lr_scheduler.OneCycleLR(
    opt,
    steps_per_epoch=10,
    pct_start=0.2,
    epochs=100,
    max_lr=1e-3,
)
print('after_init', opt.param_groups[0]['lr'])
```

The observed output is:

```text
before 0.001
after_init 0.00004
```

This shows that:

> **Even without calling `scheduler.step()` later, simply creating `OneCycleLR` is enough to change the optimizer's current learning rate from `0.001` to a much smaller starting value.**

---

## 4. Evidence 3: This Behavior Has Already Appeared in Real Experiments

In one experiment, the configuration was:

- `learning_rate = 0.001`
- `lradj = constant`

But the actual recorded behavior in the first two epochs was:

1. **Epoch 1**
   - `lr = 3.9999999999999996e-05`
   - `test_mape = 0.2460`

2. **Epoch 2**
   - `lr = 0.001`
   - `test_mape = 0.8922`

<img width="417" height="287" alt="Image" src="https://github.com/user-attachments/assets/61a30579-7d2c-4fa8-8919-148c117cf03b" />

This indicates that:

1. The effective learning rate in Epoch 1 was **not** `0.001`
2. It was actually about `4e-05`
3. It was then reset to `0.001` afterward
4. The test performance deteriorated substantially after that reset

So we already have direct experimental evidence that:

> **the script says `constant`, but the actual Epoch 1 learning rate is around `4e-05`, not `0.001`.**

---

## 5. What This Implies for CPGRU and other models ？

If CPGRU exhibits a pattern such as:

- relatively good `test_mape` in Epoch 1
- then deterioration to around `0.8` in later epochs

this is more consistent with:

> **an optimization trajectory problem, rather than a fundamentally broken model architecture**

In other words:

1. The model is not completely failing from the start
2. The effective learning-rate behavior differs between the early and later stages of training
3. Training may start in a relatively mild regime, but later become too aggressive

---

## 6. Most Plausible Current Interpretation

Based on both the code inspection and the observed experiments, my current interpretation is:

1. The `CPGRU` model itself is **not** the primary suspect
2. There is no clear evidence that the wrong dataset is being loaded
3. One major technical issue is that:

> **the implementation of `lradj=constant` in `BatteryLife-main` is not truly constant; instead, it behaves like a pseudo-constant schedule that is first influenced by `OneCycleLR` and then reset at the end of the epoch**

This is sufficient to alter the optimization trajectory in centralized training, and it is likely one of the major reasons why the current centralized CPGRU result is unstable or inconsistent with expectations.

---

## 7. Summary

> In `BatteryLife-main`, the parameter `lradj=constant` is specified in `train_eval_scripts/CPGRU.sh`, but during execution `run_main.py` still creates `OneCycleLR`, and `utils/tools.py` later resets the learning rate to `args.learning_rate` at the end of the epoch. As a result, the learning-rate schedule is not truly constant, but rather a pseudo-constant schedule influenced by `OneCycleLR`. We have already observed direct evidence of this behavior in centralized experiments, and it is likely one of the key reasons for the abnormal centralized CPGRU performance.

**Comment by Ruifeng-Tan** [comment]
Dear @oneUTer,

Thank you for pointing this out and for your recognition of our work.

I have reviewed the code and fixed this issue by adding explicit branches for different learning-rate adjustment strategies in `run_main.py` (lines 256–268):

```python
if args.lradj == 'COS':
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
        model_optim, T_max=20, eta_min=1e-8
    )
elif args.lradj == 'TST':
    scheduler = lr_scheduler.OneCycleLR(
        optimizer=model_optim,
        steps_per_epoch=train_steps,
        pct_start=args.pct_start,
        epochs=args.train_epochs,
        max_lr=args.learning_rate
    )
else:
    # constant learning rate
    scheduler = None
```
With this update, OneCycleLR will no longer be initialized when args.lradj is set to constant, so the learning rate will remain fixed as expected.

Sorry for the inconvenience, and thanks again for reporting this issue.

Best regards,
Ruifeng
